torch/_inductor/decomposition.py

```diff
@@ -243,85 +251,105 @@ def should_pad_bench(mat1, mat2, op, input=None):
+        # tradeoff between performance improvement from shape padding and overhead from addtional memory ops
         # TODO: Build a learned model which would be better than this heuristic
-        if op is torch.ops.aten.mm or op is torch.ops.aten.addmm:
-            return ori_time > pad_time * 1.3
-        else:
-            return ori_time > pad_time * 2
+        return ori_time > pad_time * 1.1
 
 
 @register_decomposition([aten.mm])
 def mm_decomp(mat1, mat2):
     if (
         config.shape_padding
-        and check_device_dtype(mat1, mat2)
+        and check_device(mat1, mat2)
         and should_pad_bench(mat1, mat2, torch.ops.aten.mm)
     ):
-        m_padded_length = get_padded_length(mat1.shape[0])
-        k_padded_length = get_padded_length(mat1.shape[1])
-        n_padded_length = get_padded_length(mat2.shape[1])
+        m_padded_length = get_padded_length(mat1.shape[0], get_alignment_size(mat1))
+        k_padded_length = get_padded_length(mat1.shape[1], get_alignment_size(mat1))
+        n_padded_length = get_padded_length(mat2.shape[1], get_alignment_size(mat2))
 
-        if k_padded_length != 0:
-            mat1 = pad_dim(mat1, k_padded_length, 1)
-            mat2 = pad_dim(mat2, k_padded_length, 0)
-            return torch.ops.aten.mm(mat1, mat2)
-        elif m_padded_length != 0:
-            mat1 = pad_dim(mat1, m_padded_length, 0)
-            return torch.ops.aten.mm(mat1, mat2)[:-m_padded_length, :]
-        elif n_padded_length != 0:
-            mat2 = pad_dim(mat2, n_padded_length, 1)
-            return torch.ops.aten.mm(mat1, mat2)[:, :-n_padded_length]
+        if m_padded_length != 0 or k_padded_length != 0 or n_padded_length != 0:
+            return pad_mm(mat1, mat2, m_padded_length, k_padded_length, n_padded_length)
 
     return NotImplemented  # go directly to lowering
 
 
+def pad_mm(mat1, mat2, m_padded_length, k_padded_length, n_padded_length):
+    if k_padded_length != 0:
+        mat1 = pad_dim(mat1, k_padded_length, 1)
+        mat2 = pad_dim(mat2, k_padded_length, 0)
+        return torch.ops.aten.mm(mat1, mat2)
+    elif n_padded_length != 0:
+        mat2 = pad_dim(mat2, n_padded_length, 1)
+        return torch.ops.aten.mm(mat1, mat2)[:, :-n_padded_length]
+    else:
+        mat1 = pad_dim(mat1, m_padded_length, 0)
+        return torch.ops.aten.mm(mat1, mat2)[:-m_padded_length, :]
+
+
 @register_decomposition([aten.bmm])
 def bmm_decomp(mat1, mat2):
     if (
         config.shape_padding
-        and check_device_dtype(mat1, mat2)
+        and check_device(mat1, mat2)
         and should_pad_bench(mat1, mat2, torch.ops.aten.bmm)
     ):
-        m_padded_length = get_padded_length(mat1.shape[1])
-        k_padded_length = get_padded_length(mat1.shape[2])
-        n_padded_length = get_padded_length(mat2.shape[2])
-
+        k_padded_length = get_padded_length(mat1.shape[2], get_alignment_size(mat1))
```

Comment:
In multihead attention with odd sequence length, seq len dimension is m and n, and padding actually helps a lot, can we leave m and n padding? Where are you seeing `.contiguous` call with padded m and n?

Reply:
Great to see shape padding helps multihead attention a lot! We observed it in two internal production models. 

Probably we could add one flag to turn off/on padding m and n dims of bmm for internal models, and turn on by default for multihead attention when enabling shape padding?

Reply:
When we tested it in 128 GPUs with delay-allreduce, there was an OOM issue due to padding m and n of bmm for two internal models, though the OOM issue does not happen in single GPU

Reply:
I kind of persist in being skeptical that the IMA we saw in the 128-gpu case was really caused by OOMs here.  Even if it was, we should leave some kind of a TODO or whatever to make it so we can turn this padding on (maybe with more memory analysis?) since we're kind of leaving perf on the table.

Reply:
Thanks @bertmaher! I will follow up on this with more analysis, and will further remove the flag once the OOM issue is addressed. Will keep you updated